bindings/rascal/utils/filter_utils.py

```diff
@@ -0,0 +1,86 @@
+import numpy as np
+
+# utility functions for Filters
+
+
+def get_index_mappings_sample_per_species(managers, sps):
```

Comment:
are the new comments sufficient to understand the indexing ?

Reply:
Unfortunately, inline comments aren't nearly as helpful at indicating the purpose of a piece of code as are clear summary comments (docstrings, in this case) for each function or large block of code. If it's hard to explain the purpose of a chunk of code in plain English, then that might be a sign that some refactoring is needed, because clearly structured code doesn't tend to require complicated explanations... but that's a task for another time.

For now, I don't think the inline comments add much; I guessed as much from the variable names and what they're doing. What I'm having trouble with is seeing the forest through the trees: Why is this function needed? What are the index conventions it's mapping between? (Could this be made into a class for neater encapsulation of the mapping?) How would someeone who's unfamiliar with the CURFilter and FPSFilter implementations use these functions/mappings in new code?

Reply:
The goal of these functions is in the name and the details are in inline code. Their purpose is completely conditioned by what happens in CURFilter and FPSFilter so I would just copy paste either of the two to implement a 'new' filter.

Reply:
Then how about mentioning that dependency in the docstring? Look, there's a reason we have docstrings, and it's to explain the function's purpose beyond the actual function name. Trying to use the function name for this purpose leads to ridiculously long function names like the one below. Instead of using inline comments, I would have described the return values of the function _in the docstring_ so that it's accessible both to automated parsing tools and to someone skimming the code.

And if these functions are tied to CURFilter or FPSFilter, why not make them members of that class? Alternatively, if they encode functionality common to both CURFilter and FPSFilter, why couldn't one use that same functionality to implement a new filter?